rfcs/47-better-error-handing.md

```diff
@@ -0,0 +1,209 @@
+	case "NoSuchKey", "NotFound":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)
+	case "AccessDenied":
+		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, err)
+	// What if we add a new case, but users already handle it themselves?
+	}
+
+	return err
+}
+```
+
+Although we declared in `11-error-handling` that unexpected errors could be changed or disappeared while dependence upgraded and no changelog for them, we'd better simply eliminate the possibility of using them.
+
+#### Expected Errors
+
+`11-error-handling` is not detailed enough. It does not specify how to define and use expected errors. Current practice has too many cases, and has some problems. Both users and implementors may be confused which kind of error should be returned at which place.
+
+- Ad-hoc string errors are not user-friendly. The user cannot use `errors.Is` & `errors.As` to handle such errors specially. We should avoid using them.
+
+- Another layer of error `struct`s may be confusing.
+  - It may be mistaken for top-level errors.
+  - Their `Err` fields are also confusing.
+    - Will they wrap yet another layer of error `struct`s? 
+    - If it will be the same for all instances, not intended for the caller to fill in, and only accessed via `Unwrap`, we can actually hide the field.
+    - The wrapped errors seem like the super class. It may be more reasonable to wrap other errors instead of being wrapped. But the order does not matter in the error chain.
+
+## Proposal
+
+So I propose the following error handling specification as a supplement of `11-error-handling`:
+
+1. Public APIs SHOULD return top-level errors, which MUST be defined as below:
+	```go
+	type SomeError struct {
+		Op  string
+		Err error
+
+		ContextA string
+		ContextB structB
+		...
+	}
+	```
+2. For the wrapped error:
+   - If it is an unexpected error, it MUST be `fmt.Errorf("%v", err)`. SDK errors SHOULD not be wrapped.
+   - If it is an expected error, it MUST be created by either
+     - `fmt.Errorf("%w: %v", SomeError, err)` where
+       - `SomeError` is defined as a sentinel error `var SomeError = errors.New("what happened")`
+       - `err` is the original SDK error
+     - `NewSomeError(contextA, contextB)` where 
+       - `SomeError` is returned and it is defined as:
+			```go
+			type SomeError struct {
+				ContextA string
+				ContextB structB
+				...
+			}
+			```
+       - `SomeError` SHOULD implement `Error`
+       - `SomeError` CAN implement `Unwrap`, returning the category (or the label) of the error, which is a sentinel error
+
+![](https://user-images.githubusercontent.com/37948597/116872935-6dd1fb00-ac49-11eb-8f25-8983d564ba66.png)
+
+In short, compared with current practice, there are three changes:
+1. Ad-hoc string errors SHOULD not be used.
+2. SDK errors SHOULD not be wrapped.
+3. Wrapped error `struct`s won't wrap another layer of error `struct`s, and its wrapped `Err` is hidden.
+
+Besides, we should list all available errors in the docs.
```

Comment:
The errors list should be introduced in this RFC or another one?

Reply:
In the website doc page?

Reply:
In this RFC maybe?

Like global pairs, all errors in go-storage should be defined via the RFC procedure.